tools/congestion-model/src/strategy/soft_backpressure.rs

```diff
@@ -0,0 +1,210 @@
+use crate::model::ChunkExecutionContext;
+use crate::strategy::QueueFactory;
+use crate::{GGas, ShardId, GAS_LIMIT, TGAS, TX_GAS_LIMIT};
+use std::collections::{BTreeMap, HashMap};
+
+#[derive(Default)]
+pub struct SoftBackpressure {
+    shard_id: Option<ShardId>,
+    all_shards: Vec<ShardId>,
+}
+
+#[derive(Default, Clone, Debug)]
+struct ShardState {
+    orig_rec_txn_input_gas: HashMap<(ShardId, ShardId), GGas>,
+    shard_current_load: GGas,
+}
+
+const MAX_DESIRED_QUEUE_SIZE: GGas = 4000 * TGAS;
+
+impl crate::CongestionStrategy for SoftBackpressure {
+    fn init(&mut self, id: ShardId, all_shards: &[ShardId], _queue_factory: &mut dyn QueueFactory) {
+        self.shard_id = Some(id);
+        self.all_shards = all_shards.to_vec();
+    }
+
+    fn compute_chunk(&mut self, ctx: &mut ChunkExecutionContext) {
+        let prev_state = self
+            .all_shards
+            .iter()
+            .map(|shard_id| {
+                (
+                    *shard_id,
+                    ctx.prev_block_info()
+                        .get(shard_id)
+                        .and_then(|block_info| block_info.get::<ShardState>().cloned())
+                        .unwrap_or_default(),
+                )
+            })
+            .collect::<BTreeMap<_, _>>();
+        let mut bucket_limits = calculate_tx_limits(&prev_state, &self.all_shards);
+        // if self.shard_id == Some(self.all_shards[0]) {
+        //     println!("Tx bucket limits: {:?}", bucket_limits);
+        // }
+
+        let mut dropped_txs = Vec::new();
+        while ctx.gas_burnt() < TX_GAS_LIMIT {
+            if let Some(tx) = ctx.incoming_transactions().pop_front() {
+                let sender = ctx.tx_sender(tx);
+                let receiver = ctx.tx_receiver(tx);
+                let gas = ctx.tx_attached_gas(tx);
+                let bucket = (sender, receiver);
+                let limit = bucket_limits.get_mut(&bucket).unwrap();
+                if gas <= *limit {
+                    let receipt = ctx.accept_transaction(tx);
+                    ctx.forward_receipt(receipt);
+                    *limit -= gas;
+                } else {
+                    dropped_txs.push(tx);
+                }
+            } else {
+                break;
+            }
+        }
+        for tx in dropped_txs.drain(..).rev() {
+            ctx.incoming_transactions().push_front(tx);
+        }
+
+        let mut orig_rec_txn_input_gas = HashMap::new();
+        let orig_tx_ids = ctx
+            .incoming_receipts()
+            .iter()
+            .map(|receipt| receipt.transaction_id())
+            .collect::<Vec<_>>();
+        for tx in orig_tx_ids {
+            let tx_sender = ctx.tx_sender(tx);
```

Comment:
Yeah, I originally was thinking of "which shard forwarded this receipt to us" but gravitated towards going all the way to the source, as that makes more sense in this strategy where the only thing we can do is limit the source. I agree that it can be useful in other strategies as well. One caveat is that it's not very easy to define the exact behavior we want here. What I've implemented here is an approximation; the more accurate technique would be to start from each txn and look at the aggregate computation it induces in each shard, but that requires a lot more bookkeeping, and since txns take different lengths in time, it's not clear how to define a recent window to use for estimation.